src/file-manager/fm-desktop-wayland-bg-dialog.c

```diff
@@ -0,0 +1,414 @@
+
+   Copyright (C) 2024 Luke <lukefromdc@hushmail.com>
+
+   The Mate Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Library General Public License as
+   published by the Free Software Foundation; either version 2 of the
+   License, or (at your option) any later version.
+
+   The Mate Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Library General Public License for more details.
+
+   You should have received a copy of the GNU Library General Public
+   License along with the Mate Library; see the file COPYING.LIB.  If not,
+   write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
+   Boston, MA 02110-1301, USA.
+
+   Author: <lukefromdc@hushmail.com>
+*/
+
+#include <config.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <X11/Xatom.h>
+#include <gtk/gtk.h>
+#include <gdk/gdkx.h>
+#include <glib/gi18n.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#include <eel/eel-glib-extensions.h>
+#include <eel/eel-gtk-extensions.h>
+#include <eel/eel-vfs-extensions.h>
+
+#include <libcaja-private/caja-directory-background.h>
+#include <libcaja-private/caja-global-preferences.h>
+
+#include "fm-desktop-icon-view.h"
+#include "fm-desktop-wayland-bg-dialog.h"
+
+
+#ifdef HAVE_WAYLAND
+
+GSettings   *background_settings;
+GtkWidget   *filebutton;
+GtkWidget   *preview;
+GtkWidget   *preview_image;
+const gchar *filename;
+GdkRGBA      color1, color2;
+const gchar  *primary_color_str,  *secondary_color_str, *shading_type;
+
+static void
+update_preview()
+{
+    static GtkCssProvider *provider;
+    gchar           *css;
+    GString         *string;
+
+    if (!provider)
+        provider =  gtk_css_provider_new ();
+
+    /*Build a color preview using a cssprovider due to requirement to handle RBGA values*/
+    string = g_string_new(NULL);
+    g_string_append (string, "#caja-wayland-bg-preview {");
+
+    if (strcmp (shading_type, "vertical-gradient") == 0)
+        g_string_append (string, "background-image: linear-gradient(to bottom,");
+
+    else if  (strcmp (shading_type, "horizontal-gradient") == 0)
+        g_string_append (string, "background-image: linear-gradient(to right,");
+
+    else
+    {
+        g_string_append (string, "background-color:");
+        g_string_append (string, primary_color_str);
+    }
+
+    if ((strcmp (shading_type, "vertical-gradient") == 0) ||
+       (strcmp (shading_type, "horizontal-gradient") == 0))
+    {
+        g_string_append (string, primary_color_str);
+        g_string_append (string, ",");
+        g_string_append (string, secondary_color_str);
+        g_string_append (string, ");");
+    }
+        g_string_append (string, "}");
+
+    css = g_string_free (string, FALSE);
```

Comment:
Memory leak

Reply:
In the code as it now stands, this is fixed: In two places so we can set a color/gradient preview or clear it from an image preview, we have after building the string

```
    css = g_string_free
    gtk_css_provider_load_from_data (provider, css, -1, NULL);

    g_free (css);
```

Reply:
We are declaring (as of now) the provider as static so that when a user repeatedly changes backgrounds, the same style context can be reached and written to when `update_preview()` runs. If I do not use static I get warnings about uninitialized use, and previewing will fail on changing backgrounds. Changing the code now to ensure its only added once as well